walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -40,19 +41,69 @@
     ArrayList<Event> sortedEvents = new ArrayList<>(events);
     Collections.sort(sortedEvents, Event.ORDER_BY_START);
 
+    Collection<Event> optionalEvents = new ArrayList<>();
+
+    ListIterator<Event> eventsIterator = sortedEvents.listIterator();
+    while (eventsIterator.hasNext()) {
+      Event event = eventsIterator.next();
+      
+      // Separates mandatory attendee events and optional attendee events into two lists and
+      // removes events that are attended by people not invited to the meeting. 
+      if (Collections.disjoint(event.getAttendees(), request.getAttendees())) {
+        // Events with only optional attendees are added to optionalEvents. 
+        if (!Collections.disjoint(event.getAttendees(), request.getOptionalAttendees())) {
+          optionalEvents.add(event);
+        }
+        // Events that don't have any mandatory attendees are removed from sortedEvents.
+        eventsIterator.remove();
+      }
+    }
+
+    Collection<TimeRange> mandatoryTimes = findAvailableTimes(sortedEvents, request);
+    Collection<TimeRange> optionalTimes = findAvailableTimes(optionalEvents, request);
+
+    // If there's no mandatory attendees, only consider optional attendees.
+    if (sortedEvents.size() == 0) {
+      return optionalTimes;
+    }
+
+    Collection<TimeRange> times = new ArrayList<>();
+
+    // Find common available times between mandatory and optional attendees. 
+    for (TimeRange mandatoryTime : mandatoryTimes) {
+      for (TimeRange optionalTime : optionalTimes) {
+        if (mandatoryTime.contains(optionalTime)) {
+          times.add(optionalTime);
+        } else if (optionalTime.contains(mandatoryTime)) {
+          times.add(mandatoryTime);
+        } else if (mandatoryTime.overlaps(optionalTime)) {
+          int start = Math.max(mandatoryTime.start(), optionalTime.start());
+          int end = Math.min(mandatoryTime.end(), optionalTime.end());
+
+          TimeRange time = TimeRange.fromStartEnd(start, end, false);
+          if (time.duration() >= request.getDuration()) {
+            times.add(time);
+          }
+        }
+      }
+    }
+
+    // Return time slots where both mandatory and optional attendees are available, if any.
+    if (times.size() > 0) {
+      return times;
+    }
```

Comment:
Put this in a helper method that returns the list of overlapping times. After all these helper methods, you would have a main method with a sequence of method calls that almost reads like high level English instructions

```
(1) get individual mandatory availability
(2) get individual optional availability
(3) get list of times that all mandatory can attend at the same time
(4) get list of times that all optional can attend at the same time
(5) get overlaps between (3) and (4)
```